transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala

```diff
@@ -429,20 +419,110 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
     )
     val linksMetadataFilenameAndMsg = jsonToFilenameAndMsg("links")(readableFiles)
 
-    val validatedFilenameAndMsg = validateJson(linksMetadataFilenameAndMsg)
+    val validatedFilenameAndMsg = validateJson(linksMetadataFilenameAndMsg, s"$inputPrefix/links")
 
     // then convert json to msg and get the filename
     val processedData = validatedFilenameAndMsg
       .withName(s"Pre-process links metadata")
-      .map {
+      .flatMap {
         case (filename, metadata) =>
-          transformLinksFileMetadata(filename, metadata)
+          transformLinksFileMetadata(filename, metadata, inputPrefix)
       }
     // then write to storage
     StorageIO.writeJsonLists(
       processedData,
       "links",
-      s"${outputPrefix}/metadata/links"
+      s"$outputPrefix/metadata/links"
     )
   }
+
+  def validateJson(
+    filenamesAndMsg: SCollection[(String, Msg)],
+    inputPrefix: String
+  ): SCollection[(String, Msg)] = {
+    validateJsonInternal(inputPrefix)(filenamesAndMsg).withName("Validate: Log Validation").map {
+      case Some(error) =>
+        error.log
+      case None =>
+    }
+    filenamesAndMsg
+  }
+
+  def validateJsonInternal(inputPrefix: String)(
+    filenamesAndMsg: SCollection[(String, Msg)]
+  ): SCollection[Option[SchemaValidationError]] = {
+    // pull out the url for where the schema definition is for each file
+    val content = filenamesAndMsg
+      .withName("Validate: Schema Definition URL")
+      .map {
+        case (filename, msg) =>
+          (
+            msg
+              .tryRead[String]("describedBy")
+              .fold[Option[String]] {
+                MissingPropertyError(
+                  s"$inputPrefix/$filename",
+                  "File has no describedBy property."
+                ).log
+                None
+              }(Some(_)),
+            (filename, msg)
+          )
+      }
+      .collect { case (Some(url), rest) => (url, rest) }
```

Comment:
There must be a way to express this using `filter` but I couldn't get it it work. The code is doing `map(.. function that creates optional elements)` and then `collect(... function that removes optional elements)`. So in theory you should be able to combine both into a `filter`..

Reply:
I think we probably want to do a similar thing as your suggestion for the crc32c; I can filter and log upstream before it even gets here

Reply:
Yeah, and I realized that it did need both map and collect (or map and filter) so there wasn't a better way to do it here.